### ISSUE TITLE
[Feature Request]

### ISSUE THREAD

**Issue by absolution1** [Ryujinx/Ryujinx#4595]
### Overview

Implement XInput API

### Smaller details

NA

### Nature of request

NA

### Why would this feature be useful?

This would be massively helpful for anyone that uses multiple xbox controllers, or uses a single xbox controller that is sometimes connected by wire or  wireless. Using SDL, wire vs wireless xbox pad connections carry separate IDs and so need reconfiguring in Ryujinx every time you change connection method.

**Comment by Squall-Leonhart** [comment]
This is unnecessary,  SDL2 already uses the superior WGI api on Windows.

**Comment by absolution1** [comment]
I think there is a well-defined use case for XInput that SDL cannot provide, and that is swappable convenience of pads that can use XInput.

With XInput, I would only need to set up players 1 to 4 once and then not touch the settings again.
I would then be free to use my wired xbox one controller, wireless xbox one controller, wired 360 controller or wireless 360 controller as player 1 without needing to touch ryujinx's input settings.

I cannot do this with SDL2.  SDL2 has a rigid structure for recognising pads which would mean that I would need to open and change Ryujinx's input settings every time I wanted to use a different pad.

**Comment by Squall-Leonhart** [comment]
Sorry, but your reasoning has no merit.

SDL via WGI is already making use of the total feature set of an Xbox controller as WGI implements full support for Xinput class devices, it actually exposes more capabilities for newer Xbox controls than the Xinput api can and ever will (Haptic triggers for instance, though not relevant to this emulator)

The auto-matic reloading of controller profiles when when they have been off prior to starting ryujinx has nothing to do with SDL itself but with the implementation of the input manager only loading profiles associated to active device.

Further more, Xinput itself has less rigity for controller id association than SDL itself, having [forever had an issue where the api may decide that the controller is on slot 2-4 following a cycle of the controller going to sleep or being hotplugged.
](https://steamcommunity.com/app/353370/discussions/1/1489992713702706615/)

**Comment by absolution1** [comment]
Yes I understand that WG exposes more features from an xbox pad but my argument is purely related to how pads connected via XInput take up slots on the host, and that is how emulators register them.

So, SDL can expose controllers in a similar slot-based way?  That sounds great; I've just never encountered this in any emulator (not just Ryujinx).  All of the emus I've experienced tie player1/2/3's registered pad to the pad's ID.  Is there a reason why this is usually the case? (Or have I been unlucky with which emus I've looked at).

So, if tying a SDL2 pad connection to a slot rather than ID is indeed possible, then I need to submit a different feature request to change the input manager?

**Comment by Squall-Leonhart** [comment]
> So, SDL can expose controllers in a similar slot-based way? That sounds great; I've just never encountered this in any emulator (not just Ryujinx). All of the emus I've experienced tie player1/2/3's registered pad to the pad's ID. Is there a reason why this is usually the case? (Or have I been unlucky with which emus I've looked at).

wx/sdljoy.h has an example of using the Joy Index.

**Comment by absolution1** [comment]
>There are a problems with interfacing an input manager on top of SDL raw.

I think I'm confused now.  So, there would be issues with relating players to pad indices?

**Comment by Epsilon99s** [comment]
> This is unnecessary, SDL2 already uses the superior WGI api on Windows.

Maybe it is superior but it doesn't register inputs from my Xbox One controller. Its GUID is 030000005e040000e002000000007801. Controllers whose GUID ends with 7801 might not be seen by emus and games that support SDL only. Or rather they can be seen but not register any inputs. There is a whole discussion about controllers with 7801 suffix: https://github.com/mdqinc/SDL_GameControllerDB/issues/478

**Comment by Squall-Leonhart** [comment]
the issue you linked is solved, its left open for documentatio nsake.
